src/QLPlan.actor.cpp

```diff
@@ -687,38 +687,58 @@ ACTOR static Future<Void> doNonIsolatedRO(PlanCheckpoint* outerCheckpoint,
                                           Reference<MetadataManager> mm) {
 	if (!dtr)
 		dtr = self->newTransaction();
-	state double startt = now();
 	state Reference<PlanCheckpoint> innerCheckpoint(new PlanCheckpoint);
 	state int nTransactions = 1;
 	state int64_t nResults = 0;
 	state FlowLock* outerLock = outerCheckpoint->getDocumentFinishedLock();
+	state Deque<std::pair<Reference<ScanReturnedContext>, Future<Void>>> bufferedDocs;
+
 	try {
 		state uint64_t metadataVersion = wait(cx->bindCollectionContext(dtr)->getMetadataVersion());
 		loop {
 			state FutureStream<Reference<ScanReturnedContext>> docs = subPlan->execute(innerCheckpoint.getPtr(), dtr);
 			state FlowLock* innerLock = innerCheckpoint->getDocumentFinishedLock();
 			state bool first = true;
-			state Future<Void> timeout = delay(3.0);
+			state Future<Void> timeout = delay(3.0, TaskMaxPriority);
+			state bool finished = false;
 
-			loop choose {
-				when(state Reference<ScanReturnedContext> doc = waitNext(docs)) {
-					// throws end_of_stream when totally finished
-					Void _ = wait(outerLock->take());
```

Comment:
Would it be equivalent and simpler to just do `Void _ = wait( outerLock->take() || timeout );` instead of splitting this into two functions?

Reply:
If I am thinking right, it would look like this

```
			loop choose {
				when(state Reference<ScanReturnedContext> doc = waitNext(docs)) {
					// throws end_of_stream when totally finished
					Void _ = wait(outerLock->take() || timeout);
					if (!timeout.isReady()) {
                                              .....
					}
				}
				when(Void _ = wait(timeout)) { break; }
			}

                        // squeeze remaining documents from the stream

                       // checkpoint planner
```

This will need special handling with `!timeout.isReady()`.

And due to the way query planner checkpoint scheme works, it's important there are no documents pending in document stream when the checkpoint is being called, that would be below this loop. I guess I can have extra code before the checkpoint to squeeze the documents left in the stream.

If so, probably having 3 `when` blocks probably clearer code.